src/v/datalake/proto_to_arrow_struct.h

```diff
@@ -0,0 +1,147 @@
+      const google::protobuf::Descriptor* message_descriptor) {
+        using namespace detail;
+        namespace pb = google::protobuf;
+
+        // Set up child arrays
+        for (int field_idx = 0; field_idx < message_descriptor->field_count();
+             field_idx++) {
+            auto desc = message_descriptor->field(field_idx);
+            if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_INT32) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::Int32Type>>());
+            } else if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_INT64) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::Int64Type>>());
+            } else if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_BOOL) {
+                _child_arrays.push_back(
+                  std::make_unique<
+                    proto_to_arrow_scalar<arrow::BooleanType>>());
+            } else if (desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_FLOAT) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::FloatType>>());
+            } else if (
+              desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_DOUBLE) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::DoubleType>>());
+            } else if (
+              desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_STRING) {
+                _child_arrays.push_back(
+                  std::make_unique<proto_to_arrow_scalar<arrow::StringType>>());
+            } else if (
+              desc->cpp_type() == pb::FieldDescriptor::CPPTYPE_MESSAGE) {
+                auto field_message_descriptor = desc->message_type();
+                if (field_message_descriptor == nullptr) {
+                    throw datalake::initialization_error(
+                      std::string("Can't find schema for nested type : ")
+                      + desc->cpp_type_name());
+                }
+                _child_arrays.push_back(std::make_unique<proto_to_arrow_struct>(
+                  field_message_descriptor));
+            } else {
+                throw datalake::initialization_error(
+                  std::string("Unknown type: ") + desc->cpp_type_name());
+            }
+        }
+        // Make Arrow data types
+
+        // This could be combined into a single loop with the one above, but
+        // this seems more readable to me
+        for (int field_idx = 0; field_idx < message_descriptor->field_count();
+             field_idx++) {
+            auto field_desc = message_descriptor->field(field_idx);
+            auto field_name = field_desc->name();
+            _fields.push_back(_child_arrays[field_idx]->field(field_name));
+        }
+        _arrow_data_type = arrow::struct_(_fields);
+
+        // Make builder
+        std::vector<std::shared_ptr<arrow::ArrayBuilder>> child_builders;
```

Comment:
Add `child_builders.reserve(_child_arrays.size())` to avoid resizing (can also apply this to `push_back()` loops above). We may also want to use `fragmented_vector` instead of `vector` to avoid oversized seastar allocs.